doc/man/skaffari.8.in
.TH Skaffari 8 "2017-05-17" "Skaffari @skaffari_VERSION@"

.SH "NAME"
Skaffari - Web based email account managemant

.SH "DESCRIPTION"
.B Skaffari
is a web based interface to manage email accounts. It is currently based on the following components that play together:
.I Cyrus-IMAP
server,
.I Postfix
MTA,
.I MySQL/MariaDB
database server,
.I Cyrus-SASL
authentication daemon and
.IR "pam_mysql".
It is inpired by the PHP web-cyradm web interface and can import the settings and database from web-cyradm for an easy migration to
.BR "Skaffari".

Using administrators and domain managers,
.B Skaffari
provides a simple administrator user management where administrators can change everything and domain managers are only responsible for their domains. Where administrators can create new domains to manage, domain managers can only change a limited set of options for their domains. Creating new accounts for a domain is restricted by per domain limitations that can be set by the administrators. A domain manager can only create and edit accounts that are inside domains the manager is responsible for. At least one administrator account is needed to run
.BR "Skaffari",
domain managers can be connected to multiple domains.

Multiple domains can be managed, each with its own set of managers and accounts. Domains can be limited by administrators to not exceed a defined amount of accounts and/or disk quota. Single accounts can have individual storage quotas (needed, if there is a domain quota) as well as individual access to IMAP, POP, Sieve and SMTP (has to be also configured in
.IR "pam_mysql").
Also every account can have multiple email addresses and forwards defined.

.B Skaffari
itself is written in Qt/C++ using the
.I Cutelyst
web application framework.

.SH "SETUP"
.B skaffaricmd(8)
has to be used to setup a
.B Skaffari
installation. It will create the configuration and database layout. For details about the setup processs, see
.BR "man 8 skaffaricmd".

.SH "FILES"
.B @CMAKE_INSTALL_SYSCONFDIR@/skaffari.ini
.RS 4
default configuration file (can be changed when starting
.IR "cutelyst-wsgi")
.RE

.SH "SEE ALSO"
.BR "skaffaricmd(8)", " skaffari.ini(5)"


CHANGELOG
------------------------------------------------------------------
Skaffari 0.0.17 - 2017-08-18 10:28:00 UTC

* domains can now be grouped in a parent/children relationship where
  accounts in the parent domain automatically get also base email
  addresses for the child domains.
* fixed the JS code in the default template to check whole domains if
  there is only a single account in the domain
* update default template to Bootstrap 4 Beta
* delete all email addresses of a domain when deleting the domain. As
  accounts can have multiple email addresses and can also have addresses
  from other domains.

------------------------------------------------------------------
Skaffari 0.0.16 - 2017-07-06 16:52:00 UTC

* New support for the status column that can be used by pam_mysql
  to determine if an account or the password has been expired.
  To make this possible, accounts gained a new column to store the
  expiration date of the password. The account check route and
  skaffaricmd --update-account-status can be used to update the
  status column.
* The account list is now sortable.
* Default template: accounts in the account list will be updated
  if they have been checked and if the check changed somthing.

------------------------------------------------------------------
Skaffari 0.0.15 - 2017-07-01 19:20:00 UTC

* domain table is now sortable (#13)
* account list now loads asynchrous and an loads more content
  on behalf of the user (#11)
* account list can be filtered by username, email addres or
  forward address
* templates can have configuration files to set some Skaffari
  behavior (#17)

------------------------------------------------------------------
Skaffari 0.0.14 - 2017-06-30 07:01:00 UTC

* use human readable sizes to set quota (#6)
* add more log file output to account management (#10)
* implement catch all functionality (#4)
* use unsigned 64bit integer to store quota values,
  fixes issue (#15)

------------------------------------------------------------------
Skaffari 0.0.13 - 2017-06-27 12:51:00 UTC

* fix deleting email address after renaming dores not work (#1)
* more logging in admin management (#3)
* use AJAX for more actions in default template (#2)

------------------------------------------------------------------
Skaffari 0.0.12 - 2017-06-23 15:55:00 UTC

* make email management routes AJAX friendly
* use AJAX for email address management
* add localization support

------------------------------------------------------------------
Skaffari 0.0.11 - 2017-06-19 09:32:00 UTC

* store email addresses in ACE format
* fix gettings single domain
* use QCollator for locale aware sorting
* add HelpEntry class to unify help entries in templates
* add general skaffari 8 man page
* move help text into core instead of template files
* begin improving input template forms
* add functions to check account consistency

------------------------------------------------------------------
Skaffari 0.0.10 - 2017-05-11 11:44:00 UTC

* IMAP: implement mailbox management (create, delete, subscribe, setacl,
  setquota)
* fix some issues in template

------------------------------------------------------------------
Skaffari 0.0.9 - 2017-05-09 20:26:00 UTC

* fixing fucking typos

------------------------------------------------------------------
Skaffari 0.0.8 - 2017-05-09 20:11:00 UTC

* fix wrong key name in map

------------------------------------------------------------------
Skaffari 0.0.7 - 2017-05-09 19:40:00 UTC

* fix man page names

------------------------------------------------------------------
Skaffari 0.0.6 - 2017-05-09 19:27:00 UTC

* improve IMAP response parsing
* rename some configuration key names and variables to be not confused
* set default values for configuration in main CMakeLists.txt file to
  use them in man pages
* rename skaffarictl to skaffaricmd

------------------------------------------------------------------
Skaffari 0.0.5 - 2017-05-07 23:53:00 UTC

* add new SkaffariConfig class as central configuration storage and use
  it all over Skaffari
* fix MySQL initialization when using mutliple threads, workaround for
  QTBUG-54872
* SkaffariIMAP: remove overladoed constructors and use SkaffariConfig
  in default constructor
* define configuration default values in config.h file
* skaffarictl: improve IMAP parsing (has to be applied to SkaffariIMAP)
* add option to set a different peer name for SLL/TLS verification
* get rid of SkaffariEngine and SEngine, both are not needed anymore

------------------------------------------------------------------
Skaffari 0.0.4 - 2017-05-04 22:53:00 UTC

* improve packaging
* improve startup error reporting
* add testing switch to skaffarictl
* add skaffarictl man page

------------------------------------------------------------------
Skaffari 0.0.3 - 2017-05-02 13:21:00 UTC

* Fixed missing installation of skaffarictl

------------------------------------------------------------------
Skaffari 0.0.2 - 2017-05-02 13:13:00 UTC

* fix missing CMakeLists.txt files in release tarball

------------------------------------------------------------------
Skaffari 0.0.1 - 2017-05-02 13:03:00 UTC

* early preview release with basic functionality



src/skaffari.cpp
/*
 * Skaffari - a mail account administration web interface based on Cutelyst
 * Copyright (C) 2017 Matthias Fehring <kontakt@buschmann23.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "skaffari.h"

#include <Cutelyst/Application>
#include <Cutelyst/Plugins/StaticSimple/StaticSimple>
#include <Cutelyst/Plugins/View/Grantlee/grantleeview.h>
#include <Cutelyst/Plugins/Session/Session>
#include <Cutelyst/Plugins/Authentication/authentication.h>
#include <Cutelyst/Plugins/Authentication/credentialpassword.h>
#include <Cutelyst/Plugins/Authentication/authenticationrealm.h>
#include <Cutelyst/Plugins/Utils/Sql>
#include <Cutelyst/Plugins/StatusMessage>
#include <Cutelyst/Plugins/Memcached/Memcached>
#include <Cutelyst/Plugins/MemcachedSessionStore/MemcachedSessionStore>
#include <Cutelyst/Engine>
#include <Cutelyst/Plugins/CSRFProtection/CSRFProtection>
#include <grantlee5/grantlee/metatype.h>
#include <grantlee5/grantlee/engine.h>
#include <QSqlDatabase>
#include <QSqlError>
#include <QDir>
#include <QMetaType>
#include <QCoreApplication>
#include <QTranslator>
#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QLoggingCategory>

extern "C"
{
#ifdef WITH_SYSTEMD
#define SD_JOURNAL_SUPPRESS_LOCATION
#include <systemd/sd-journal.h>
#endif

#include <syslog.h>
}

#include "objects/domain.h"
#include "objects/simpleadmin.h"
#include "objects/simpledomain.h"
#include "objects/simpleaccount.h"
#include "objects/adminaccount.h"
#include "objects/account.h"
#include "objects/folder.h"
#include "objects/helpentry.h"
#include "objects/emailaddress.h"
#include "objects/skaffarierror.h"

#include "utils/language.h"
#include "utils/skaffariconfig.h"

#include "../common/config.h"
#include "../common/global.h"
#include "root.h"
#include "authstoresql.h"
#include "login.h"
#include "logout.h"
#include "domaineditor.h"
#include "accounteditor.h"
#include "admineditor.h"
#include "myaccount.h"
#include "settingseditor.h"

Q_LOGGING_CATEGORY(SK_CORE, "skaffari.core")

using namespace Cutelyst;

bool Skaffari::isInitialized = false;
bool Skaffari::messageHandlerInstalled = false;

#ifdef WITH_SYSTEMD
void journaldMessageOutput(QtMsgType type, const QMessageLogContext &context, const QString &msg)
{
    int prio = LOG_INFO;
    switch (type) {
    case QtDebugMsg:
        prio = LOG_DEBUG;
        break;
    case QtInfoMsg:
        prio = LOG_INFO;
        break;
    case QtWarningMsg:
        prio = LOG_WARNING;
        break;
    case QtCriticalMsg:
        prio = LOG_CRIT;
        break;
    case QtFatalMsg:
        prio = LOG_ALERT;
        break;
    }

#ifdef QT_DEBUG
    sd_journal_send("PRIORITY=%i", prio, "SYSLOG_FACILITY=1", "SYSLOG_IDENTIFIER=%s", context.category, "SYSLOG_PID=%lli", QCoreApplication::applicationPid(), "MESSAGE=%s", qFormatLogMessage(type, context, msg).toUtf8().constData(), "CODE_FILE=%s", context.file, "CODE_LINE=%s", context.line, "CODE_FUNC=%s", context.function, NULL);
#else
    sd_journal_send("PRIORITY=%i", prio, "SYSLOG_FACILITY=1", "SYSLOG_IDENTIFIER=%s", context.category, "SYSLOG_PID=%lli", QCoreApplication::applicationPid(), "MESSAGE=%s", qFormatLogMessage(type, context, msg).toUtf8().constData(), NULL);
#endif

    if (prio == 0) {
        abort();
    }
}
#endif

void syslogMessageOutput(QtMsgType type, const QMessageLogContext &context, const QString &msg)
{
    int prio = LOG_INFO;
    switch (type) {
    case QtDebugMsg:
        prio = LOG_DEBUG;
        break;
    case QtInfoMsg:
        prio = LOG_INFO;
        break;
    case QtWarningMsg:
        prio = LOG_WARNING;
        break;
    case QtCriticalMsg:
        prio = LOG_CRIT;
        break;
    case QtFatalMsg:
        prio = LOG_ALERT;
        break;
    }

    openlog(context.category, LOG_PID, LOG_USER);
    syslog(prio, "%s", qFormatLogMessage(type, context, msg).toUtf8().constData());
    closelog();

    if (prio == 0) {
        abort();
    }
}

Skaffari::Skaffari(QObject *parent) : Application(parent)
{
}

Skaffari::~Skaffari()
{
}

bool Skaffari::init()
{
    const QVariantMap generalConfig = engine()->config(QStringLiteral("Skaffari"));

    if (!messageHandlerInstalled) {
        const QString backend = generalConfig.value(QStringLiteral("logging_backend")).toString();
        if (backend.compare(QLatin1String("syslog"), Qt::CaseInsensitive) == 0) {
            qSetMessagePattern(QStringLiteral("%{message}"));
            qInstallMessageHandler(syslogMessageOutput);
            qCInfo(SK_CORE, "Using syslog logging backend.");
        }
#ifdef WITH_SYSTEMD
        else if (backend.compare(QLatin1String("journald"), Qt::CaseInsensitive) == 0) {
            qSetMessagePattern(QStringLiteral("%{message}"));
            qInstallMessageHandler(journaldMessageOutput);
            qCInfo(SK_CORE, "Using systemd's journald logging backend.");
        }
#endif
        else {
            qCInfo(SK_CORE, "Using stdout logging backend.");
        }
        messageHandlerInstalled = true;
    }

    QCoreApplication::setApplicationName(QStringLiteral("Skaffari"));
    QCoreApplication::setApplicationVersion(QStringLiteral(SKAFFARI_VERSION));

    qRegisterMetaType<quota_size_t>("quota_size_t");
    qRegisterMetaType<dbid_t>("dbid_t");
    qRegisterMetaType<Folder>();
    qRegisterMetaType<Domain>();
    qRegisterMetaType<SimpleAdmin>();
    qRegisterMetaType<SimpleDomain>();
    qRegisterMetaType<SimpleAccount>();
    qRegisterMetaType<AdminAccount>();
    qRegisterMetaType<Language>();
    qRegisterMetaType<Account>();
    qRegisterMetaType<HelpEntry>();
    qRegisterMetaType<HelpHash>("HelpHash");
    qRegisterMetaType<EmailAddress>();
    qRegisterMetaType<SkaffariError>();

    Grantlee::registerMetaType<Folder>();
    Grantlee::registerMetaType<Domain>();
    Grantlee::registerMetaType<SimpleAdmin>();
    Grantlee::registerMetaType<SimpleDomain>();
    Grantlee::registerMetaType<SimpleAccount>();
    Grantlee::registerMetaType<AdminAccount>();
    Grantlee::registerMetaType<Language>();
    Grantlee::registerMetaType<Account>();
    Grantlee::registerMetaType<HelpEntry>();
    Grantlee::registerMetaType<EmailAddress>();

    const QString tmplName = generalConfig.value(QStringLiteral("template"), QStringLiteral("default")).toString();
    const QString tmplBasePath = QStringLiteral(SKAFFARI_TMPLDIR) + QLatin1Char('/') + tmplName;
    SkaffariConfig::setTmplBasePath(tmplBasePath);

    if (!isInitialized) {
        QVariantMap tmplConfig;
        QFile tmplConfigFile(tmplBasePath + QLatin1String("/metadata.json"));
        if (tmplConfigFile.exists()) {
            qCDebug(SK_CORE, "Found template metadata file.");
            if (tmplConfigFile.open(QIODevice::ReadOnly|QIODevice::Text)) {
                QJsonParseError jpe;
                QJsonDocument tmplJsonConfig(QJsonDocument::fromJson(tmplConfigFile.readAll(), &jpe));
                if (jpe.error != QJsonParseError::NoError) {
                    qCCritical(SK_CORE, "Failed to parse template metadata file: %s", qUtf8Printable(jpe.errorString()));
                    return false;
                }

                tmplConfig = tmplJsonConfig.object().value(QStringLiteral("config")).toObject().toVariantMap();
            } else {
                qCCritical(SK_CORE, "Failed to open template metadata file %s.", qUtf8Printable(tmplConfigFile.fileName()));
                return false;
            }
        }

        qCDebug(SK_CORE) << "Initializing configuration.";
        SkaffariConfig::load(generalConfig,
                             engine()->config(QStringLiteral("Accounts")),
                             engine()->config(QStringLiteral("Admins")),
                             engine()->config(QStringLiteral("IMAP")),
                             tmplConfig);

        if (SkaffariConfig::imapUser().isEmpty()) {
            qCCritical(SK_CORE) << "No valid IMAP user defined.";
            return false;
        }

        if (SkaffariConfig::imapPassword().isEmpty()) {
            qCCritical(SK_CORE) << "No valid IMAP password defined.";
            return false;
        }

        // initialize DB one time to prevent https://bugreports.qt.io/browse/QTBUG-54872
        if (!initDb()) {
            return false;
        }

        SkaffariConfig::loadSettingsFromDB();

        isInitialized = true;
    }

    QString sitePath = tmplBasePath + QLatin1String("/site");

    auto view = new GrantleeView(this);
    view->setTemplateExtension(QStringLiteral(".html"));
    view->setWrapper(QStringLiteral("wrapper.html"));
    view->setCache(false);
    view->setIncludePaths({sitePath});
    view->engine()->addDefaultLibrary(QStringLiteral("grantlee_i18ntags"));
    view->engine()->addDefaultLibrary(QStringLiteral("grantlee_skaffari"));

    /* Start loading translations */
    const QString tmplTransFileName = QLatin1String("tmpl_") + tmplName;
    const QString tmplTransFilePath = tmplBasePath + QLatin1String("/l10n");

    for (const QString &lang : SKAFFARI_SUPPORTED_LANGS) {
        if (Q_LIKELY(lang != QLatin1String("en"))) {
            qCDebug(SK_CORE, "Loading translations for language %s.", qUtf8Printable(lang));
            const QLocale locale(lang);
            if (Q_LIKELY(locale.language() != QLocale::C)) {
                auto coreTrans = new QTranslator(this);
                if (Q_LIKELY(coreTrans->load(locale, QStringLiteral("skaffari"), QStringLiteral("_"), QStringLiteral(SKAFFARI_L10NDIR)))) {
                    addTranslator(locale, coreTrans);
                } else {
                    qCWarning(SK_CORE, "Failed to load core translation file for language %s from %s.", qUtf8Printable(locale.bcp47Name()), SKAFFARI_L10NDIR);
                    delete coreTrans;
                }

                auto tmplTrans = new QTranslator(this);
                if (Q_LIKELY(tmplTrans->load(locale, tmplTransFileName, QStringLiteral("_"), tmplTransFilePath))) {
                    view->addTranslator(locale, tmplTrans);
                } else {
                    qCWarning(SK_CORE, "Failed to load template translation file for language %s from %s.", qUtf8Printable(locale.bcp47Name()), qUtf8Printable(tmplTransFilePath));
                    delete tmplTrans;
                }
            } else {
                qCWarning(SK_CORE, "Invalid language code: %s", qUtf8Printable(lang));
            }
        }
    }
    /* End loading translations */

    new Root(this);
    new Login(this);
    new Logout(this);
    new DomainEditor(this);
    new AccountEditor(this);
    new AdminEditor(this);
    new MyAccount(this);
    new SettingsEditor(this);

    auto staticSimple = new StaticSimple(this);
    const QString staticPath = tmplBasePath + QLatin1String("/static");
    staticSimple->setIncludePaths({staticPath, QStringLiteral(SKAFFARI_STATICDIR)});

    if (SkaffariConfig::useMemcached()) {
        auto memc = new Memcached(this);
        memc->setDefaultConfig({
                                   {QStringLiteral("binary_protocol"), true}
                               });
    }

    auto sess = new Session(this);

    if (SkaffariConfig::useMemcachedSession()) {
        sess->setStorage(new MemcachedSessionStore(this, this));
    }

    auto csrf = new CSRFProtection(this);
    csrf->setDefaultDetachTo(QStringLiteral("/csrfdenied"));

    new StatusMessage(this);

    auto auth = new Authentication(this);
    auto cred = new CredentialPassword;
    cred->setPasswordType(CredentialPassword::Hashed);
    auto store = new AuthStoreSql;
    auth->addRealm(store, cred);

    defaultHeaders().setHeader(QStringLiteral("X-Frame-Options"), QStringLiteral("DENY"));
    defaultHeaders().setHeader(QStringLiteral("X-Content-Type-Options"), QStringLiteral("nosniff"));
    defaultHeaders().setHeader(QStringLiteral("X-XSS-Protection"), QStringLiteral("1; mode=block"));
    defaultHeaders().setHeader(QStringLiteral("Content-Security-Policy"), QStringLiteral("default-src 'none'; script-src 'self'; style-src 'self'; font-src 'self'; img-src 'self' data:; connect-src 'self'; base-uri 'self'; form-action 'self';"));
    defaultHeaders().setHeader(QStringLiteral("X-Robots-Tag"), QStringLiteral("none"));
    defaultHeaders().setHeader(QStringLiteral("Referrer-Policy"), QStringLiteral("no-referrer, strict-origin-when-cross-origin"));

    return true;
}


bool Skaffari::postFork()
{

    initDb();

    return true;
}

bool Skaffari::initDb() const
{
    const QVariantMap dbconfig = engine()->config(QStringLiteral("Database"));
    const QString dbtype = dbconfig.value(QStringLiteral("type")).toString();
    const QString dbname = dbconfig.value(QStringLiteral("name")).toString();
    const QString dbuser = dbconfig.value(QStringLiteral("user")).toString();
    const QString dbpass = dbconfig.value(QStringLiteral("password")).toString();
    const QString dbhost = dbconfig.value(QStringLiteral("host"), QStringLiteral("localhost")).toString();
    const int dbport = dbconfig.value(QStringLiteral("port"), QStringLiteral("3306")).toInt();

    qCDebug(SK_CORE) << "Establishing database connection";
    QSqlDatabase db;
    const QString dbConName = Sql::databaseNameThread();
    if (dbtype == QLatin1String("QMYSQL")) {
        if (dbname.isEmpty()) {
            qCCritical(SK_CORE) << "No database name set!";
            return false;
        }
        if (dbuser.isEmpty()) {
            qCCritical(SK_CORE) << "No database user set!";
            return false;
        }
        if (dbpass.isEmpty()) {
            qCCritical(SK_CORE) << "No database password set!";
            return false;
        }

        db = QSqlDatabase::addDatabase(dbtype, dbConName);
        if (Q_LIKELY(db.isValid())) {
            db.setDatabaseName(dbname);
            db.setUserName(dbuser);
            db.setPassword(dbpass);

            if (dbhost[0] == QLatin1Char('/')) {
                db.setConnectOptions(QStringLiteral("UNIX_SOCKET=%1;MYSQL_OPT_RECONNECT=1;CLIENT_INTERACTIVE=1").arg(dbhost));
            } else {
                db.setConnectOptions(QStringLiteral("MYSQL_OPT_RECONNECT=1;CLIENT_INTERACTIVE=1"));
                db.setHostName(dbhost);
                db.setPort(dbport);
            }
        } else {
            qCCritical(SK_CORE) << "Failed to obtain database object.";
            return false;
        }
    } else {
        qCCritical(SK_CORE) << dbtype << "is not a supported database type.";
        return false;
    }

    if (Q_UNLIKELY(!db.open())) {
        qCCritical(SK_CORE) << "Failed to establish database connection:" << db.lastError().text();
        return false;
    }

    return true;
}

#include "moc_skaffari.cpp"
